### DISCUSSION TITLE
fallback: true - fallback page could optionally load ASAP when next.js Link is clicked

### DISCUSSION THREAD

**Discussion by r00dY** [vercel/next.js#23992] [Ideas]
# Feature request

Scenario.
1. We have a page `/product/[handle]` with `fallback: true`.
2. Let's say `/product/123` is not yet generated statically.
3. We have a page `/collection` with next link to `/product/123`, no prefetching, or haven't prefetched yet.
4. We click on a link.
5. Nothing is happening until page is generated and then page reloads. Fallback page wasn't loaded at any point in time.

Last point could have optional alternative behaviour: showing fallback page ASAP.

## Is your feature request related to a problem? Please describe.

We already designed a loading page for when `router.isFallback = true`. I want to reuse this loader for next.js reloads when static site is being generated.

Now after clicking next.js link I have either 1-2 seconds moment of "nothing happening" which is bad UX. I cold build a loader on router events, but then I'd have to maintain two loading states instead of one.

## Describe the solution you'd like

A flag for `Link`? Something like `showFallbackImmediately`? Could default to `false` to preserve backward compatibility. 

Potentially it could be a trait of a page too.

**Comment by Visakeswaran** [comment]
**next 9.0.2**
I am facing a similar issue. I am routing to a dynamic route having around 12 api calls inside getInitialProps(), using next/link. when i trigger routing, the page becomes unresponsive until all the api fetching is done. I have referred to the with-loading example and also disabled prefetch for link, but no good.

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/17028188/89517172-faed3e00-d7f6-11ea-8ea6-873f51fe8708.gif)

**Comment by fgnass** [comment]
+1 for configuring the behavior at page-level rather than the Link. The same pattern should be applied (after opt-in) while the client waits for the `getServerSideProps` to be fetched.

**Comment by yunyong** [comment]
I have same issue.
```
 if (router.isFallback) {
    return <div>Loading...</div>
}
```
'Loading...' does not appear in **PRODUCTION** mode. (in development, it works)
When I click on the link, nothing happens for 1~2 sec.
I want to change the page ASAP and show a spinner for users.

**Comment by danielcolinjames** [comment]
I actually thought this was how `fallback: true` already worked, but just implemented it and realized it doesn't work like that... Did anyone figure out any way to achieve this? I really wanted to eliminate that 1-2 seconds of no-visual-feedback loading time when clicking a Next `<Link />` and `getStaticProps` has to build the page...

**Comment by nathanaelphilip** [comment]
I too would like this. Seems like an essential feature when page loads take a few seconds.

**Comment by danny-dang** [comment]
In my case, `router.isFallBack` doesn't work really well for me. The page still freeze for about few seconds before redirecting to a new route.
After several hours of reading docs. I found these 2 events from router: `routeChangeStart` and `routeChangeComplete`
Basically you can listen to when the page has finish redirecting.

I'm doing something like this
```javascript
  const router = useRouter()
...

  useEffect(() => {
    const handleRouteChange = (url) => {
      dispatch(setLoading(true))
      console.log(
        `App is changing to ${url} `
      )
    }

    const handleRouteComplete = (url) => {
      dispatch(setLoading(false))
      console.log(
        `App changed to ${url} `
      )
    }

    router.events.on('routeChangeStart', handleRouteChange)
    router.events.on('routeChangeComplete', handleRouteComplete)

    return () => {
      router.events.off('routeChangeStart', handleRouteChange)
      router.events.off('routeChangeComplete', handleRouteComplete)
    }
  }, [])
  
  if(isLoading){
     return <Loading/>
  }
 ...
```

**Reply by leephan2k1** [reply]
Thanks a lot, it saved my time. Next JS 12 still has this problem. NextJS should do this behind the scenes Link instead of doing it manually like this.
I don't understand what they recommend RFC for when this problem persists.

**Comment by meandillar** [comment]
> In my case, `router.isFallBack` doesn't work really well for me. The page still freeze for about few seconds before redirecting to a new route.
> After several hours of reading docs. I found these 2 events from router: `routeChangeStart` and `routeChangeComplete`...

I've tried using the router events as suggested above but I have the same problem as using `router.isFallback`. The `routeChangeStart` event doesn't fire on initial navigation so I still get a delay while the page builds.

Seems like `router.isFallback` works as expected if you go directly to a url and allows you to display a loader, but not when navigating using a next/link.

EDIT: Solution above using router events actually works great. I initially implemented it in the wrong place.

EDIT2: We can't use the router events solution because it unmounts the page on shallow navigate which means we lose local state. For example when update a query string.

**Comment by LOLsho** [comment]
I have the same issue.
It would be very nice to render page right after clicking on a link and show loader while data for page is loading

**Reply by AlexThunders** [reply]
+1

**Comment by steven-tey** [comment]
Any updates on this issue? @danny-dang's workaround did work but it feels like a temporary fix – the route still shows the old path even when the DOM changes. Would love if this could be baked into both next/Link and router.push functions! 💯

**Comment by berndartmueller** [comment]
I would also be in favor of this. I initially thought that's how `isFallback: true` works generally. After deploying to Vercel I immediately experienced this few seconds delay before transitioning to the fallback page.

**Reply by goellner** [reply]
i had the same assumption, thats not clear from the docs and it really throws users off, when nothing happens at all for a second or two

**Reply by SILVAWesley** [reply]
Same! I was trying to figure out why my 'fallback: true' wasn't working when I found this discussion topic...

**Reply by 0mjs** [reply]
Hm, seems I've stumbled along the same thing. I had an SSR page structure with good use of `getServerSideProps` to hit the API Routes Next.js  provides (also using next-connect) for better Express sugar. Anyways, it was super slow, like mentioned above, from there I started down the rabbit-hole that is CSR/SSR/SSG/ISR and this ongoing effort to implement something that works for every pattern. After implementing a solely SWR approach, fetching data from my API which calls some MongoDB query functions, I realised that I _could_ load the skeletons/loading well, and it _did_ route directly there, the UX was awesome, it even seemed to cache well. I soon realised that ASO (https://nextjs.org/docs/advanced-features/automatic-static-optimization) was being used by default, as those pages don't use either of the provided static or SSR functions. Now, I've gone to an ISR approach which statically generates the data and updates it as needed, albeit rendering stale data for some time. But, the routing is the same as SSR? Even with fallback and conditionally using the isFallback param, the page still renders as it did in an SSR manner. I'm confused and feel this is the right place to chime in.

**Comment by tyuen** [comment]
One workaround is to use regular `<a href>` links instead of `<link href>` so that `isFallback` can be used.

**Reply by noahsaso** [reply]
This is such a simple and elegant solution. Can't believe I didn't think of it, thank you 🙏

**Reply by MartinDavi** [reply]
Yeah but you're unloading and reloading the entire site - you lose all the benefit of client side routing.

**Comment by luigi-derson** [comment]
In my case they work in development but not in production after build. 

The build process generates some of all dynamic routes we have because there is thousands of routes (product catalogue). 

The pages that are not generated at build time and use `router.isFallback()` are being correctly loaded if you access them by putting the URL in the browser's address bar on go, like `/p/1` or if you refresh the page. But if you use Link component to go to this page from the homepage `<Link href="/p/1" />` it won't load that page data. It just shows a page without any props until you refresh it.

**Comment by carlosrsabreu** [comment]
Are there any updates about this? 

The `router.isFallback()` only works when we request the page directly and when it has no prefetching / hasn't prefetched yet. But when we get the page through a redirect from `router.push()` or `next/link`, it doesn't work (with` <a>` we can get it...).

**Comment by alexandereneroth** [comment]
Why is this a feature request and not a bug, what am i missing?
The way i understand the docs this should already be happening, no?

**Comment by sspenst** [comment]
Really don't understand why this hasn't been fixed or addressed yet. `Link` is one of the fundamental components in Next.js, so it seems fair to assume users will arrive at most pages via a `Link`.

Switching to `a` links as @tyuen mentioned will end up showing the `router.isFallback` content, but then you get rid of all prefetching and client-side route transitions between pages which seems worse overall.

Currently it feels like `router.isFallback` is essentially useless until this is fixed.

**Reply by sebastianpikand** [reply]
Can you elaborate please why does switching to <a> seems worse overall? Do I understand correctly that if the page has not been statically built (i.e. no visits), having <a> tag gives the loading state, but then the next visitors requesting the already generated page suffer in user experience, because data is not prefetched?

In summary:
<a> – better exp. for first-time visit, worse exp for subsequent visits
<Link> – awful exp. for first-time visits, great exp for subsequent visits (bc of prefetching)

**Comment by delbaoliveira** [comment]
Hi everyone, 

Incase you missed it, we recently shared an RFC that includes new routing behavior: https://nextjs.org/blog/layouts-rfc

There is likely some crossover with the points raised here and the [conclusion](https://nextjs.org/blog/layouts-rfc#conclusion) where we discuss Instant Loading States (and possibly Offscreen Stashing with Instant Back/Forward).

>Instant Loading States: With server-side routing, rendering and data fetching happens on the server before navigation completes. This makes it important to show loading UI so the application doesn't feel unresponsive. We propose framework level support for instant loading states - inline and global loading indicators and skeletons.

>Offscreen Stashing with Instant Back/Forward: React is planning to add a new <Offscreen /> component that stores a React tree and its state without rendering it to the screen. Leveraging this component, it should be possible to stash routes that have been visited and pre-render routes before they are visited. Navigation backwards and forwards to these routes should be instant and restore any previously stored state.

We welcome any feedback here: https://github.com/vercel/next.js/discussions/37136

**Comment by qruz-hq** [comment]
It's been 2 years since this issue was opened and it's still hanging around :/ This really decreases UX imo. Hopefully, it gets patched soon 🙏

**Comment by charlesoestergaard** [comment]
Still an issuse here for me aswell....

**Comment by sebastianpikand** [comment]
Reading between the lines it seems that for loading state to work we have to use /app directory with Next.js 13... which I am doing with one new project, but not with an older one :/
